include/iceflow/executor.hpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2024 The IceFlow Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ICEFLOW_NODE_EXECUTOR_HPP
+#define ICEFLOW_NODE_EXECUTOR_HPP
+
+#include "iceflow.hpp"
+
+#include <iceflow.grpc.pb.h>
+
+#include <grpc/grpc.h>
+
+namespace iceflow {
+
+class ExternalExecutor {
+public:
+  virtual ~ExternalExecutor() {}
+  virtual void receiveCongestionReport(CongestionReason congestionReason) = 0;
+};
+
+class IceflowExecutor : public std::enable_shared_from_this<IceflowExecutor> {
+public:
+  IceflowExecutor(const std::string &serverAddress,
+                  const std::string &clientAddress,
+                  std::shared_ptr<ExternalExecutor> externalExecutor);
+
+  ~IceflowExecutor();
+
+  void receiveCongestionReport(CongestionReason congestionReason);
```

Comment:
Ideally, the congestion report would also include information regarding the edge that is congested, although it is not strictly necessary.

_Usually_, we would detect that an edge in the graph is congested based on the difference between produced and consumed units summed over all nodes.
However, this only works based on the assumption that all nodes are equal and that there is no network congestion.
In any other case, only some of the instances of a given task might be overloaded, and in those cases, it might be good to explicitly provide information about the congested edge to the node executor.

Reply:
Looks good for now, although I am unsure if we really want to define a name (in string form) for each edge separately.

Reply:
Hmm, do you think a numeric identifier would make more sense here then?

Reply:
Or do you mean that we should make the field/parameter optional?

Reply:
Ideally, we'd have some identifier that can be automatically generated based on the start and end node of the DAG (so that we don't need to generate more identifiers), but that might cause problems for some graphs (e.g. if there are multiple edges between the same two nodes).

I'd keep it as a string for now, maybe we can change that later though.